mtt/src/mtt_types.rs

```diff
@@ -118,3 +116,226 @@ impl MttL1EntryType {
+    }
+}
+
+fn get_l1_index_and_subindex(phys_addr: SupervisorPageAddr) -> (u64, u64) {
+    let addr_bits = phys_addr.bits();
+    // Bits 25:17 for the 9-bit index, then 3-bit left shift for the 8-byte entry
+    let l1_index = (addr_bits & 0x03fe_0000) >> 14;
+    // 5-bit sub-index (16:12) with 1-bit left shift for the 2-bits per entry
+    let sub_index = (addr_bits & 0x1_f000) >> 11;
+    (l1_index, sub_index)
+}
+
+fn get_l1_value(pfn: u64, l1_index: u64) -> u64 {
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *const u64;
+    assert!(l1_index <= 4096 - 8);
+    // Safety: The 4K L1 page is allocated by the TSM, and we are using a 12-bit index.
+    unsafe { *l1_entry_addr }
+}
+
+// Helper function to update value for an existing L1 entry
+fn update_l1_value(pfn: u64, phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let (l1_index, sub_index) = get_l1_index_and_subindex(phys_addr);
+    let value = get_l1_value(pfn, l1_index);
+    // Clear the original 2-bits and OR in the new value
+    let value = (value & !(3 << sub_index)) | ((memory_type as u64) << sub_index);
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *mut u64;
+    assert!(l1_index <= 4096 - 8);
+    // Safety: The 4K L1 page is allocated by the TSM, and we are using a 12-bit index.
+    unsafe {
+        *l1_entry_addr = value;
+    };
+}
+
+// The L1 is a 4KB page that's allocated by the TSM to manage the memory type (non-confidential or confidential) at a 4K-page granularity.
+// This is done by dividing the 64MB region containing the page into 16KB regions of 4K each, and the 2-bit encoding in the L1 page pointed
+// to by the L2 entry determines the type.
+// The 12-bit index into the 4K page is computed using the 14-bits [25:12] of the physical address, and then shifting right by 2
+fn update_l1_entry(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let addr = phys_addr.bits();
+    let l2_entry = get_l2_entry(phys_addr);
+    match l2_entry {
+        MttL1Dir(pfn) => update_l1_value(pfn, phys_addr, memory_type),
+        _ => panic!("Attempt to set MttL1Dir on {l2_entry:?} for {addr:x}"),
+    }
+}
+
+/// Changes the L2 type for 16 consecutive entries (each representing a 64MB range) for the 1GB range
+/// containing the address to the specified type.
+/// This can be called to consolidate previously fractured ranges.
+pub fn set_1gb_range_type(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let gb_range_start = PageAddr::new(SupervisorPhysAddr::supervisor(
+        PageSize::Size1G.round_down(phys_addr.bits()),
+    ))
+    .unwrap();
+    for i in 0u64..=15 {
+        // Unwrap OK: We are starting with an aligned address and are adding 0 or 16K 4K pages (64MB)
+        let range_addr = gb_range_start.checked_add_pages(i * 16 * 1024).unwrap();
+        // Sanity check
+        let _ = get_l2_entry(range_addr);
```

Comment:
If this is fallible it should return an error.

Also consider an `entry_mut` helper so both get and set don't need to be called?

Reply:
Ack.

Reply:
The latest refactoring should address the above...PTAL.